File: xterminal_device.c
#include <mongoose.h>
#include <pty.h>
#include <net/if.h>
#include "common.h"

struct session {
	pid_t pid;
	int pty;
	char id[33];
	struct list_head node;
	ev_io watcher;
	char topic_dev[128];
	char topic_srv[128];
	struct mg_connection *nc;
};

static char *devid;
static LIST_HEAD(session_list);

static void pty_read_cb(struct ev_loop *loop, ev_io *w, int revents)
{
	int len = 0;
	char buf[1024] = "";
	struct session *s = (struct session *)w->data;
	
	len = read(w->fd, buf, sizeof(buf));
	if (len > 0) {
		mg_mqtt_publish(s->nc, s->topic_dev, 0, MG_MQTT_QOS(0), buf, len);
		printf("pub to:%s %s\n", s->topic_dev, buf);
	} else {
		perror("read");
		ev_io_stop(loop, w);
	}
}

static struct session *find_session_by_id(char *id)
{
	struct session *s = NULL;
	
	list_for_each_entry(s, &session_list, node) {
		if (!strcmp(s->id, id))
			return s;
	}
	
	return NULL;
}

static void new_connect(struct mg_connection *nc, char *id)
{
	struct session *s = NULL;
	pid_t pid;
	int pty;
	struct mg_mqtt_topic_expression topic_expr[1];
	
	pid = forkpty(&pty, NULL, NULL, NULL);
	if (pid < 0) {
		perror("forkpty");
		return;
	} else if (pid == 0) {
		execl("/bin/login", "login", NULL);
	}
	
	s = calloc(1, sizeof(struct session));
	if (!s) {
		perror("calloc");
		return;
	}
	
	list_add(&s->node, &session_list);
	strncpy(s->id, id, sizeof(s->id));
	s->pid = pid;
	s->pty = pty;
	s->nc = nc;
	
	topic_expr[0].qos = MG_MQTT_QOS(0);
	asprintf((char **)&topic_expr[0].topic, "xterminal/%s/%s/srvdata", devid, id);
	mg_mqtt_subscribe(nc, topic_expr, 1, 44);
	printf("sub:%s\n", topic_expr[0].topic);
	free((char *)topic_expr[0].topic);
	
	snprintf(s->topic_dev, sizeof(s->topic_dev), "xterminal/%s/%s/devdata", devid, id);
	
	ev_io_init(&s->watcher, pty_read_cb, pty, EV_READ);
	s->watcher.data = s;
	ev_io_start(EV_DEFAULT, &s->watcher);
}

static void ev_handler(struct mg_connection *nc, int ev, void *ev_data)
{
	if (ev == MG_EV_CONNECT) {
		/* Connect to mqtt broker */
		struct mg_send_mqtt_handshake_opts opts;
		memset(&opts, 0, sizeof(opts));

		mg_set_protocol_mqtt(nc);
		mg_send_mqtt_handshake_opt(nc, devid, opts);
		
	} else if (ev == MG_EV_MQTT_CONNACK) {
		struct mg_mqtt_message *msg = (struct mg_mqtt_message *)ev_data;
		struct mg_mqtt_topic_expression topic_expr[1];
		
		if (msg->connack_ret_code != MG_EV_MQTT_CONNACK_ACCEPTED) {
			printf("Got mqtt connection error: %d\n", msg->connack_ret_code);
			ev_break(EV_DEFAULT, EVBREAK_ALL);
			return;
		}
		
		topic_expr[0].qos = MG_MQTT_QOS(0);
		asprintf((char **)&topic_expr[0].topic, "xterminal/%s/connect/+", devid);
		
		mg_mqtt_subscribe(nc, topic_expr, 1, 0);

		free((char *)topic_expr[0].topic);
		
	} else if (ev == MG_EV_MQTT_PUBLISH) {
		struct mg_mqtt_message *msg = (struct mg_mqtt_message *)ev_data;
		
		if (memmem(msg->topic.p, msg->topic.len, "connect", strlen("connect"))) {
			char id[33] = "";
			strncpy(id, msg->topic.p + 31, 32);
			
			new_connect(nc, id);
		} else {
			char  id[33] = "";
			struct session *s = NULL;
			
			strncpy(id, msg->topic.p + 23, 32);
			s = find_session_by_id(id);
			if (s) {
				int ret = write(s->pty, msg->payload.p, msg->payload.len);
				if (ret < 0)
					perror("write");
			}
			
		}
	} else if (ev == MG_EV_CLOSE) {
		printf("connection close:%p\n", nc);
	}
}

static void signal_cb(struct ev_loop *loop, ev_signal *w, int revents)
{
	if (w->signum == SIGINT) {
		ev_break(loop, EVBREAK_ALL);
	}
}

static void heartbeat_cb(struct ev_loop *loop, ev_timer *w, int revents)
{
	struct mg_connection *nc = (struct mg_connection *)w->data;
	mg_mqtt_publish(nc, "xterminal/heartbeat", 0, MG_MQTT_QOS(0), devid, strlen(devid));
}

static int get_iface_mac(const char *ifname, char mac[6])
{
    int r, s;
    struct ifreq ifr;

    strncpy(ifr.ifr_name, ifname, 15);
    ifr.ifr_name[15] = '\0';

    s = socket(AF_INET, SOCK_DGRAM, 0);
    if (-1 == s) {
		perror("socket");
		return -1;
    }

    r = ioctl(s, SIOCGIFHWADDR, &ifr);
    if (r == -1) {
        perror("ioctl");
        close(s);
        return -1;
    }

    close(s);

	memcpy(mac, ifr.ifr_hwaddr.sa_data, 6);
	
    return 0;
}

int main(int argc, char *argv[])
{
	struct ev_loop *loop = EV_DEFAULT;
	ev_signal sigint_watcher;
	ev_timer heartbeat_timer;
	struct mg_mgr mgr;
	struct mg_connection *nc;
	char mac[6] = "";
	const char *s_address = "localhost:1883";
	
	mg_mgr_init(&mgr, NULL);

	nc = mg_connect(&mgr, s_address, ev_handler); 
	if (!nc) {
		fprintf(stderr, "mg_connect(%s) failed\n", s_address);
		exit(1);
	}

	ev_signal_init(&sigint_watcher, signal_cb, SIGINT);
	sigint_watcher.data = nc;
	ev_signal_start(loop, &sigint_watcher);
	
	ev_timer_init(&heartbeat_timer, heartbeat_cb, 0.1, 2.0);
	heartbeat_timer.data = nc;
	ev_timer_start(loop, &heartbeat_timer);

	get_iface_mac("enp3s0", mac);

	devid = calloc(1, 13);
	sprintf(devid, "%02hhX%02hhX%02hhX%02hhX%02hhX%02hhX",
		mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);

	printf("devid:%s\n", devid);
	
	ev_run(loop, 0);

	printf("exit...\n");

	free(devid);
	
	mg_mgr_free(&mgr);

	return 0;
}

File: xterminal_device.lua
#!/usr/bin/lua

local ev = require("ev")
local evmg = require("evmongoose")
local cjson = require("cjson")
local posix = evmg.posix
local loop = ev.Loop.default

local ARGV = arg
local show = false
local log_to_stderr = false
local ifname = "eth0"
local mqtt_host = "localhost"
local mqtt_port = "1883"
local devid = nil
local session = {}

local function log_init()
	local opt = posix.LOG_ODELAY
	if log_to_stderr then
		opt = opt + posix.LOG_PERROR 
	end
	posix.openlog("xterminal device", opt, posix.LOG_USER)
end

local function logger(level, ...)
	posix.syslog(level, table.concat({...}, "  "))
end

local function usage()
	print("Usage:", ARGV[0], "options")
	print([[
        -s              Only show config
        -d              Log to stderr
        -i              default is eth0
        --mqtt-port     default is 1883
        --mqtt-host 	default is localhost
	]])
	os.exit()
end

local function parse_commandline()
	local longopt = {
		{"help",  false, 'h'},
		{"mqtt-port", true, "0"},
		{"mqtt-host", true, "0"}
	}
	
	for r, optarg, longindex in posix.getopt(ARGV, "hsdi:", longopt) do
		if r == "d" then
			log_to_stderr = true
		elseif r == "i" then
			ifname = optarg
		elseif r == "s" then
			show = true
		elseif r == "0" then
			local name = longopt[longindex][1]
			if name == "mqtt-port" then
				mqtt_port = optarg
			elseif name == "mqtt-host" then
				mqtt_host = optarg
			end
		else
			usage()
		end
	end
end

local function show_conf()
	print("log_to_stderr", log_to_stderr)
	print("mqtt-port", mqtt_port)
	print("mqtt-host", mqtt_host)
	print("ifname", ifname)
	os.exit()
end

local function get_dev_id(ifname)
	local file = io.open("/sys/class/net/" .. ifname .. "/address", "r")
	if not file then return nil end
	local d = file:read("*l")
	file:close()
	
	if not d then return nil end
	
	return d:gsub(":", ""):upper()
end

local function new_connect(con, id)
	local pid, pty = posix.forkpty()
	if pid == 0 then posix.exec("/bin/login") end
	
	session[id] = {pid = pid, pty = pty}
	
	local topic = {
		{name = "xterminal/todev/data/" .. id},
		{name = "xterminal/todev/disconnect/" .. id},
		{name = "xterminal/uploadfile/" .. id}
	}
		
	con:mqtt_subscribe(topic);
	
	session[id].rio = ev.IO.new(function(loop, w, revents)
		local d, err = posix.read(w:getfd(), 1024)
		if not d then
			w:stop(loop)
			posix.wait(pid)
			session[id] = nil
			con:mqtt_publish("xterminal/touser/disconnect/" .. id, "");
			logger(posix.LOG_INFO, "session close:", id)
			return
		end
		con:mqtt_publish("xterminal/touser/data/" .. id, d);
	end, pty, ev.READ)
	
	session[id].rio:start(loop)
end

local keep_alive = 3
local alive_timer = ev.Timer.new(function(loop, w, event)
	keep_alive = keep_alive - 1
	if keep_alive == 0 then w:stop(loop) end
end, 5, 5)

local heartbeat_timer

local function ev_handle(con, event)
	local mgr = con:get_mgr()
	
	if event == evmg.MG_EV_CONNECT then
		local result = con:get_evdata()
		if not result.connected then
			logger(posix.LOG_ERR, "connect failed:", result.err)
		else
			con:mqtt_handshake({clean_session = true})
		end
		
	elseif event == evmg.MG_EV_MQTT_CONNACK then
		local msg = con:get_evdata()
		if msg.code ~= evmg.MG_EV_MQTT_CONNACK_ACCEPTED then
			logger(posix.LOG_ERR, "Got mqtt connection error:", msg.code, msg.err)
			return
		end
		
		local topic = {
			{name = "xterminal/connect/" .. devid ..  "/+"}
		}
		con:mqtt_subscribe(topic);
		heartbeat_timer = ev.Timer.new(function(loop, timer, revents) con:mqtt_publish("xterminal/heartbeat/" .. devid, "") end, 0.1, 3)
		heartbeat_timer:start(loop)
		alive_timer:start(loop)
		
		logger(posix.LOG_INFO, "connect", mqtt_host, mqtt_port, "ok")
		
	elseif event == evmg.MG_EV_MQTT_PUBLISH then
		local msg = con:get_evdata()
		local topic, payload = msg.topic, msg.payload
		
		if topic:match("xterminal/connect/%w+") then
			local id = topic:match("xterminal/connect/" .. devid .. "/(%w+)")
			logger(posix.LOG_INFO, "New connection:", id)
			new_connect(con, id)
		elseif topic:match("xterminal/todev/data/%w+") then
			local id = topic:match("xterminal/todev/data/(%w+)")
			posix.write(session[id].pty, payload)
		elseif topic:match("xterminal/todev/disconnect/%w+") then
			local id = topic:match("xterminal/todev/disconnect/(%w+)")
			if session[id] then
				logger(posix.LOG_INFO, "connection close:", id)
				posix.kill(session[id].pid, posix.SIGKILL)
				session[id].rio:stop(loop)
			end
		elseif topic:match("xterminal/uploadfile/%w+") then
			local topic, payload = con:mqtt_recv()
			local id = topic:match("xterminal/uploadfile/(%w+)")
			local url, file = payload:match("(%S+) (%S+)")
			local cmd = string.format("rm -f /tmp/%s;wget -q -P /tmp -T 5 %s/%s", file, url, file) 
			
			mgr:connect_http(function(con2, event2)
				if event2 == evmg.MG_EV_HTTP_REPLY then
					con2:set_flags(evmg.MG_F_CLOSE_IMMEDIATELY)
					local body = con2:body()
					
					local file = io.open("/tmp/" .. file, "w")
					file:write(body)
					file:close()
					
					con:mqtt_publish("xterminal/uploadfilefinish/" .. id, "");
				end
			end, string.format("%s/%s",url, file))
		end
	
	elseif event == evmg.MG_EV_MQTT_PINGRESP then
		keep_alive = 3
		
	elseif event == evmg.MG_EV_POLL then
		if keep_alive == 0 then
			-- Disconnect and reconnection
			logger(posix.LOG_ERR, "keep_alive timeout")
			con:set_flags(evmg.MG_F_CLOSE_IMMEDIATELY)
			heartbeat_timer:stop(loop)
			keep_alive = 3
		end
		
	elseif event == evmg.MG_EV_CLOSE then
		ev.Timer.new(function()
			logger(posix.LOG_ERR, "Try Reconnect...")
			mgr:connect(ev_handle, mqtt_host .. ":" .. mqtt_port)
		end, 5):start(loop)
	end
end

local function main()
	parse_commandline()
	if show then show_conf() end
	log_init()
	
	ev.Signal.new(function(loop, sig, revents)
		loop:unloop()
	end, ev.SIGINT):start(loop)

	devid = get_dev_id(ifname)
	if not devid then
		logger(posix.LOG_ERR, "get dev id failed for", ifname)
		os.exit()
	end
	logger(posix.LOG_INFO, "devid:", devid)
	
	local mgr = evmg.init()
	mgr:connect(ev_handle, mqtt_host .. ":" .. mqtt_port)
	
	loop:loop()
	logger(posix.LOG_INFO, "exit...")
	posix.closelog()
end

main()